### PR TITLE
[Cider] Flatten initial

### PR DESCRIPTION
**Pull request by EclecticGriffin**
A bunch of the initial data structures for cider 2.0. Lots unfinished but figured give people a chance to take a look at things. Don't bother looking at the translation stuff too closely as it is very unfinished and has some bugs.

### CODE REVIEW

File: interp/src/flatten/structures/index_trait.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+pub trait IndexRef: Copy + Eq {
+    fn index(&self) -> usize;
+    fn new(input: usize) -> Self;
+}
+
+/// This macro is used to implement the IndexRef trait for a type that wraps an
+/// unsigned integer value. By default, the macro will implement the trait using
+/// a [`u32`](std::u32) as the backing type. However, if a different backing type
+/// is desired, it can be specified as the second argument.
+macro_rules! impl_index {
```
Comment (rachitnigam) [inline interp/src/flatten/structures/index_trait.rs:11]: Potential alternative you might enjoy: https://github.com/cucapra/filament/blob/main/src/utils/idx.rs

File: interp/src/primitives/primitive_traits.rs [hunk 2]
```diff
@@ -29,7 +29,7 @@ pub trait Primitive: Named {
     /// Validate inputs to the component.
     fn validate(&self, inputs: &[(ir::Id, &Value)]);
 
-    /// Execute the component.ß
```
Comment (rachitnigam) [inline interp/src/primitives/primitive_traits.rs:32]: is this beta-reduction?

Comment (EclecticGriffin) [reply interp/src/primitives/primitive_traits.rs:32]: Legally I am required to say yes

File: interp/src/flatten/structures/index_trait.rs (hunk 1 above)
Comment (EclecticGriffin) [reply interp/src/flatten/structures/index_trait.rs:11]: Yeah I honestly cannot decide which I prefer. There are certainly some merits to either approach. The main issue I see for the interpreter is that we want many types of indices which themselves are not necessarily "representing" a singular type in the original IR. For example we want local port indices for definitions which themselves can be turned into global indices for the state table and both philosophically represent ports from a certain view but are not the `IR::Port` type. And it seems awkward to invent new empty types to associate with `Idx` for that purpose. Hopefully that makes some sort of sense?

**Comment by EclecticGriffin** [comment]
You can now translate single component programs into the internal flattened representations. Try:
```
cargo run -- tests/control/if.futil flat
```

This'll print out the debug representations of the primary & secondary context structs and then crash with a note about how the flattened stuff cannot do interpretation yet

**Comment by rachitnigam** [comment]
(leaving this here mostly as a note to myself) As we discussed in the meeting , it might be useful to merge this even though it is in a "partial state" as long as it doesn't break anything else

**Pull request merged**